### ISSUE TITLE
[ F# ] autocompletes are inaccurate

### ISSUE THREAD

**Issue by dsyme** [mono/monodevelop#3892]
F# auto-completions are inaccurate.  You can witness this by just typing

    let someFunction argggg = arg

in various places and see if the completion for `arggggg` appears. Sometimes it does, sometimes it doesn't, sometimes you get `no completions`.

There may be several causes for this, but one problem appears to be that stale information is frequently used when editing within the current line.  

There is [code to force a reparse](https://github.com/mono/monodevelop/blob/d4627becd4fef461a2e667ddbac41280d0d80209/main/external/fsharpbinding/MonoDevelop.FSharpBinding/FSharpTextEditorCompletion.fs#L434) but
1. it is not invoked very often, only when they user types exactly `->` and a few other such cases.  It should be reapplied more or less every time the user starts typing a fresh word
2. It uses RunSynchronously, which is wrong, as well as a lock inside an async, which is even wronger

cc @nosami

> VS bug #569960

**Comment by dsyme** [comment]
@nosami I took a glance at the code.

One option would be to remove the single-line caching, and always call ParseAndCheckFileInProject whenever new completions are requested.  This is what the Visual F# Tools do, see https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs#L101.  This gives total accuracy. assuming the IDE is smart enough to only call for completions when, for example, a new word starts.  The rest of the linked code is a good guide.

However, if we feel like we need improvements here then I suggest to  save the original line text from the parsing in FSharpParsedDocument (currently you just save the location of the caret). Then we can reuse the completions if the lines are "similar".

**Comment by nosami** [comment]
> One option would be to remove the single-line caching, and always call ParseAndCheckFileInProject whenever new completions are requested

Yeah.. that is what I was planning on doing. The call to ParseAndCheckFileInProject invocation would be called whenever the popup is displayed, which is roughly once per word. Will see how it feels.

**Comment by nosami** [comment]
It's noticeably slower to call ParseAndCheckFileInProject  for each completion.

With caching, intellisense feels snappy.

![intellisense-no-reparse](https://user-images.githubusercontent.com/667194/36588662-b3d25ca4-1880-11e8-93bc-846a36fe8498.gif)

Without caching, it feels sluggish and distracting as I need to wait for the popup.

![intellisense-reparse](https://user-images.githubusercontent.com/667194/36588677-bbb15e02-1880-11e8-8be3-e25e3b7ef19b.gif)

It would be good to identify the cases where a reparse is actually needed within a line - I don't think that there are that many.

**Comment by dsyme** [comment]
It's reasonable to seek to recycle for a continuous sequence of A.B.C and some other characters too likely, given you have it already.

I wonder if we could put this recycling logic in FCS so it could be shared.

**Comment by nosami** [comment]
> I wonder if we could put this recycling logic in FCS so it could be shared.

We could, but the completion APIs would need to change - it's currently doing this

```fs
checkResults.GetDeclarationListSymbols(...)
```

where checkResults is the thing we cache.

**Comment by nosami** [comment]
While looking into this issue, I spotted this in the FCS trace while trying to complete `DateTime` here `System.|`

```
FCS: recovering from error in GetDeclarationListSymbols: 'internal error: tcrefOfAppTy'
```

**Comment by dsyme** [comment]
We should drfinitely fix that

**Comment by nosami** [comment]
I think this fixes it. Currently testing. It only seems to error when trying to complete with stale typecheck results.

```diff
✔ ~/src/FSharp.Compiler.Service [master L|✚ 1…20⚑ 6]
10:48 $ git diff
diff --git a/src/fsharp/service/service.fs b/src/fsharp/service/service.fs
index 44a1d6b6a..d004ef533 100755
--- a/src/fsharp/service/service.fs
+++ b/src/fsharp/service/service.fs
@@ -993,7 +993,9 @@ type TypeCheckInfo
                             | Item.FakeInterfaceCtor (TType_app(tcref,_))
                             | Item.DelegateCtor (TType_app(tcref,_)) -> tcref.CompiledName
                             | Item.CtorGroup (_, (cinfo :: _)) ->
-                                (tcrefOfAppTy g cinfo.ApparentEnclosingType).CompiledName
+                                match tryDestAppTy g cinfo.ApparentEnclosingType with
+                                | Some tcref -> tcref.CompiledName
+                                | None -> d.Item.DisplayName
                             | _ -> d.Item.DisplayName)

                     // Filter out operators (and list)
```

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-28] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Pull Request #4960 merged by: Jason Imison  
 Author: mono  
 Commit: b5b9c681030f753d0025d8b3621cfbabd6636dd5 (xamarin/monodevelop)
